### PR TITLE
[SPARK-33389][SQL] Make internal classes of SparkSession always using active SQLConf

### PR DESCRIPTION
**Pull request by luluorta**
### What changes were proposed in this pull request?

This PR makes internal classes of SparkSession always using active SQLConf. We should remove all `conf: SQLConf`s from ctor-parameters of this classes (`Analyzer`, `SparkPlanner`, `SessionCatalog`, `CatalogManager` `SparkSqlParser` and etc.) and use  `SQLConf.get` instead.

### Why are the changes needed?

Code refine.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing test

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/HasConf.scala [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ */
+
+package org.apache.spark.sql.catalyst
+
+import org.apache.spark.sql.internal.SQLConf
+
+/**
+ * Trait for shared SQLConf.
+ */
+trait HasConf {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/HasConf.scala:25]: This name is weird. How about  `SQLConfHelper`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/HasConf.scala [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst
+
+import org.apache.spark.sql.internal.SQLConf
+
+/**
+ * Trait for shared SQLConf.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/HasConf.scala:23]: `Trait for getting the active SQLConf.`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 3]
```diff
@@ -42,7 +42,7 @@ import org.apache.spark.sql.catalyst.trees.TreeNodeRef
 import org.apache.spark.sql.catalyst.util.toPrettySQL
 import org.apache.spark.sql.connector.catalog._
 import org.apache.spark.sql.connector.catalog.CatalogV2Implicits._
-import org.apache.spark.sql.connector.catalog.TableChange.{AddColumn, After, ColumnChange, ColumnPosition, DeleteColumn, RenameColumn, UpdateColumnComment, UpdateColumnNullability, UpdateColumnPosition, UpdateColumnType}
+import org.apache.spark.sql.connector.catalog.TableChange.{First => _, _}
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:45]: unnecessary change?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 4]
```diff
@@ -61,34 +61,38 @@ class SessionCatalog(
...
+    cacheTTL: Long = SQLConf.get.metadataCacheTTL) extends Logging with HasConf {
   import SessionCatalog._
   import CatalogTypes.TablePartitionSpec
 
   // For testing only.
   def this(
       externalCatalog: ExternalCatalog,
       functionRegistry: FunctionRegistry,
-      conf: SQLConf) = {
+      staticConf: SQLConf) = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:76]: The previous name `conf` is OK. `SQLConf` contains both static and runtime configs.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/BooleanSimplificationSuite.scala [hunk 5]
```diff
@@ -188,27 +188,29 @@ class BooleanSimplificationSuite extends PlanTest with ExpressionEvalHelper with
...
+    new SessionCatalog(new InMemoryCatalog, EmptyFunctionRegistry))
 
   test("(a && b) || (a && c) => a && (b || c) when case insensitive") {
-    val plan = caseInsensitiveAnalyzer.execute(
-      testRelation.where(('a > 2 && 'b > 3) || ('A > 2 && 'b < 5)))
-    val actual = Optimize.execute(plan)
-    val expected = caseInsensitiveAnalyzer.execute(
-      testRelation.where('a > 2 && ('b > 3 || 'b < 5)))
-    comparePlans(actual, expected)
+    withSQLConf(SQLConf.CASE_SENSITIVE.key -> "false") {
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/BooleanSimplificationSuite.scala:195]: This is the default value. Seems we can remove `withSQLConf`?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/BooleanSimplificationSuite.scala [hunk 6]
```diff
@@ -188,27 +188,29 @@ class BooleanSimplificationSuite extends PlanTest with ExpressionEvalHelper with
...
   }
 
   test("(a || b) && (a || c) => a || (b && c) when case insensitive") {
-    val plan = caseInsensitiveAnalyzer.execute(
-      testRelation.where(('a > 2 || 'b > 3) && ('A > 2 || 'b < 5)))
-    val actual = Optimize.execute(plan)
-    val expected = caseInsensitiveAnalyzer.execute(
-      testRelation.where('a > 2 || ('b > 3 && 'b < 5)))
-    comparePlans(actual, expected)
+    withSQLConf(SQLConf.CASE_SENSITIVE.key -> "false") {
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/BooleanSimplificationSuite.scala:206]: ditto

File: sql/core/src/test/resources/sql-tests/inputs/datetime.sql [hunk 7]
```diff
@@ -149,7 +149,7 @@ select to_timestamp('2019-10-06 A', 'yyyy-MM-dd GGGGG');
 select to_timestamp('22 05 2020 Friday', 'dd MM yyyy EEEEEE');
 select to_timestamp('22 05 2020 Friday', 'dd MM yyyy EEEEE');
 select unix_timestamp('22 05 2020 Friday', 'dd MM yyyy EEEEE');
-select from_json('{"time":"26/October/2015"}', 'time Timestamp', map('timestampFormat', 'dd/MMMMM/yyyy'));
+select from_json('{"timestamp":"26/October/2015"}', 'timestamp Timestamp', map('timestampFormat', 'dd/MMMMM/yyyy'));
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/datetime.sql:152]: why do we change this test?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveTableScanExec.scala [hunk 8]
```diff
@@ -60,8 +59,6 @@ case class HiveTableScanExec(
   require(partitionPruningPred.isEmpty || relation.isPartitioned,
     "Partition pruning predicates only supported for partitioned tables.")
 
-  override def conf: SQLConf = sparkSession.sessionState.conf
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveTableScanExec.scala:63]: This actually makes sense, to make sure the conf matches the spark session. How about we move this override to `SparkPlan`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/HasConf.scala (hunk 1 above)
Comment (luluorta) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/HasConf.scala:25]: I agree, ` SQLConfHelper` is better.

File: sql/core/src/test/resources/sql-tests/inputs/datetime.sql (hunk 7 above)
Comment (luluorta) [reply sql/core/src/test/resources/sql-tests/inputs/datetime.sql:152]: After this PR, dynamically set "spark.sql.ansi.enabled" actually takes effect in parsing phase. This query will failed parsing cause `time` is a reserved key word of SQL standard.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/HasConf.scala (hunk 2 above)
Comment (luluorta) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/HasConf.scala:23]: done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 3 above)
Comment (luluorta) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:45]: reverted

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala (hunk 4 above)
Comment (luluorta) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:76]: reverted

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/BooleanSimplificationSuite.scala (hunk 5 above)
Comment (luluorta) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/BooleanSimplificationSuite.scala:195]: reverted

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/BooleanSimplificationSuite.scala (hunk 6 above)
Comment (luluorta) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/BooleanSimplificationSuite.scala:206]: reverted

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveTableScanExec.scala (hunk 8 above)
Comment (luluorta) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveTableScanExec.scala:63]: reverted

File: sql/core/src/test/resources/sql-tests/inputs/datetime.sql (hunk 7 above)
Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/inputs/datetime.sql:152]: `timestamp` is also reserved in ANSI. How about `ts`?

Comment (luluorta) [reply sql/core/src/test/resources/sql-tests/inputs/datetime.sql]: done

File: sql/core/src/test/resources/sql-tests/inputs/datetime.sql [hunk 9]
```diff
@@ -149,7 +149,7 @@ select to_timestamp('2019-10-06 A', 'yyyy-MM-dd GGGGG');
 select to_timestamp('22 05 2020 Friday', 'dd MM yyyy EEEEEE');
 select to_timestamp('22 05 2020 Friday', 'dd MM yyyy EEEEE');
 select unix_timestamp('22 05 2020 Friday', 'dd MM yyyy EEEEE');
-select from_json('{"time":"26/October/2015"}', 'time Timestamp', map('timestampFormat', 'dd/MMMMM/yyyy'));
-select from_json('{"date":"26/October/2015"}', 'date Date', map('dateFormat', 'dd/MMMMM/yyyy'));
-select from_csv('26/October/2015', 'time Timestamp', map('timestampFormat', 'dd/MMMMM/yyyy'));
-select from_csv('26/October/2015', 'date Date', map('dateFormat', 'dd/MMMMM/yyyy'));
+select from_json('{"ts":"26/October/2015"}', 'ts Timestamp', map('timestampFormat', 'dd/MMMMM/yyyy'));
```
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/inputs/datetime.sql:152]: which change caused this?

Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/inputs/datetime.sql:152]: It was a bug before that the parser used in `from_json` to parse the schema string sticks to the configs of the first created session in the current thread. Now the parser always use the active conf, and ANSI test fails here because `time` is a reserved keyword.

We probably need a separate PR for this bug.

Comment (luluorta) [reply sql/core/src/test/resources/sql-tests/inputs/datetime.sql]: I opened a new PR for this issue
[https://github.com/apache/spark/pull/30357](url)

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveTableScanExec.scala [hunk 10]
```diff
@@ -172,7 +172,7 @@ case class HiveTableScanExec(
         prunePartitions(hivePartitions)
       }
     } else {
-      if (sparkSession.sessionState.conf.metastorePartitionPruning &&
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveTableScanExec.scala:175]: let's keep this unchanged for now. We may override `def conf` in `SparkPlan` later, to always get conf from the captured spark session.

Comment (luluorta) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveTableScanExec.scala]: Reverted

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
GA passed, merging to master, thanks!
